### PR TITLE
Google/calendar list

### PR DESCRIPTION
**Pull request by Gautime**
Added the functionality to view the list of calendars present on the uesr's profile using simple command `/calendar list`. User will be able to see list of all the calendars on their profile. Along with that list, there will appear buttons with each calendar in the list. Users can click on one of those buttons and select the calendar they want to create, update, delete, read events from or to that calendar. Screenshot is attached for reference. Button will appear for all calendars except the one which is already selected ad being worked with.

Also the login and logout links are converted to a single button. Logout button example is attached in the screenshot.

Screenshot from 2019-07-16 20-49-09

### CODE REVIEW

File: helpers/persistence.ts [hunk 1]
```diff
@@ -32,4 +33,19 @@ export class AppPersistence {
...
+        await this.persistence.updateByAssociations([user_association], { calendar }, true);
+
+    }
+
+    public async get_preferred_calendar_id(user: IUser): Promise<void> {
+
+        const user_association = new RocketChatAssociationRecord(RocketChatAssociationModel.USER, user.id);
+        const [result] = await this.persistenceRead.readByAssociation(user_association);
+        const calendar_id = result ? (result as any).calendar : undefined;
+        return calendar_id;
```
[the remark this answers is not in the document]
Comment (Gautime) [reply helpers/persistence.ts]: Changed it to do something else but is now returning `result` itself. :)

File: helpers/GSGetter.ts [hunk 2]
```diff
@@ -168,6 +173,64 @@ export class GCGetter {
...
+                const list_api_response = await http.get(list_url, { headers: { 'Authorization': `Bearer ${list_token}`, } });
+                console.log('This is the calendar list respose:', list_api_response);
+
+                let all_calendars: Array<string> = list_api_response.data.items;
+
+                for (var i = 0; i < list_api_response.data.items.length; i++) {
+                    all_calendars[i] = list_api_response.data.items[i].summary;
+                }
+
+                console.log('This is calendar name inside list command:', all_calendars[2])
```
[the remark this answers is not in the document]
Comment (Gautime) [reply helpers/GSGetter.ts]: Removed the console logs as well. :)

File: helpers/GSGetter.ts [hunk 3]
```diff
@@ -58,12 +66,18 @@ export class GCGetter {
                 break;
 
             case (Command.lgout):
-
-                // const mesg = modify.getCreator().startMessage().setSender(context.getSender()).setRoom(context.getRoom());
-                const logresponse = `https://www.google.com/accounts/Logout?continue=${redirect}`;
+                const logresponse = `https://www.google.com/accounts/Logout?continue=https://appengine.google.com/_ah/logout?continue=${redirect}`;
```
Comment (d-gubert) [inline helpers/GSGetter.ts]:
```suggestion
                const logresponse = `https://www.google.com/accounts/Logout?continue=${redirect}`;
```

File: helpers/Webhook.ts [hunk 4]
```diff
@@ -1,27 +1,20 @@
...
 import { GCResults } from '../helpers/GCResult';
-import { IUser } from '@rocket.chat/apps-engine/definition/users';
 import { AppPersistence } from '../helpers/persistence';
 
-
 export class WebhookEndpoint extends ApiEndpoint {
     public path = 'webhook';
-    private readonly urli = 'http://localhost:3000/api/apps/public/c759c4f1-a3c1-4202-8238-c6868633ed87/webhook';
     public tokenid;
+    private readonly urli = 'http://localhost:3000/api/apps/public/c759c4f1-a3c1-4202-8238-c6868633ed87/webhook';
```
Comment (d-gubert) [inline helpers/Webhook.ts]: This should be dynamically retrieved by the app instead of hardcoded

File: helpers/GSGetter.ts (hunk 3 above)
Comment (Gautime) [reply helpers/GSGetter.ts]: The `redirect` variable will retrieve the site url as entered in the app settings, as for the rest of the url `https://www.google.com/accounts/Logout?continue=https://appengine.google.com/_ah/logout?continue=`, this is provided by google to log a user out and remains the same for all the users. just the site url which changes is retrieved dynamically.

Comment (d-gubert) [reply helpers/GSGetter.ts]: Sounds confusing 👀 Do you have a reference page?

Comment (Gautime) [reply helpers/GSGetter.ts]: Yes sure, it is described in one of the answers here  -
 https://stackoverflow.com/questions/12909332/how-to-logout-of-an-application-where-i-used-oauth2-to-login-with-google

**Pull request merged**